tests/unit_tests/scheduler/test_openpbs_driver.py

```diff
@@ -106,3 +107,69 @@ async def test_job_name_with_prefix():
     driver = OpenPBSDriver(job_prefix="pre_")
     await driver.submit(0, "sleep", name="sleepy")
     assert " -Npre_sleepy " in Path("captured_qsub_args").read_text(encoding="utf-8")
+
+
+@pytest.mark.parametrize(
+    ("exit_code, error_msg"),
+    [
+        (171, "Invalid credential"),
+        (183, "Premature end of message"),
+        (162, "Connection refused"),
+        (199, "Not recognized"),
+    ],
+)
+async def test_that_qsub_will_retry_and_fail(
+    monkeypatch, tmp_path, exit_code, error_msg
+):
+    os.chdir(tmp_path)
+    bin_path = tmp_path / "bin"
+    bin_path.mkdir()
+    monkeypatch.setenv("PATH", f"{bin_path}:{os.environ['PATH']}")
+    qsub_path = bin_path / "qsub"
+    qsub_path.write_text(f"#!/bin/sh\necho {error_msg} >&2\nexit {exit_code}")
+    qsub_path.chmod(qsub_path.stat().st_mode | stat.S_IEXEC)
+    driver = OpenPBSDriver(num_retries=2, retry_interval=0.2)
+    match_str = (
+        f"failed after 2 retries with error {error_msg}"
+        if exit_code != 199
+        else "failed with exit code 199 and error message: Not recognized"
+    )
+    with pytest.raises(RuntimeError, match=match_str):
+        await driver.submit(0, "sleep 10")
+
+
+@pytest.mark.parametrize(
+    ("exit_code, error_msg"),
+    [
+        (171, "Invalid credential"),
+        (183, "Premature end of message"),
+        (162, "Connection refused"),
+    ],
+)
+async def test_that_qsub_will_retry_and_succeed(
+    monkeypatch, tmp_path, exit_code, error_msg
+):
+    os.chdir(tmp_path)
+    bin_path = tmp_path / "bin"
+    bin_path.mkdir()
+    monkeypatch.setenv("PATH", f"{bin_path}:{os.environ['PATH']}")
+    qsub_path = bin_path / "qsub"
+    qsub_path.write_text(
+        "#!/bin/sh"
+        + dedent(
+            f"""
+            TRY_FILE="{bin_path}/script_try"
+            if [ -f "$TRY_FILE" ]; then
+                echo "SUCCESS"
```

Comment:
For consistency, you should probably output a mocked job id to stdout (the driver will probably use `SUCCESS` as the job id though)